src/suspendplugin.c

```diff
@@ -118,44 +121,66 @@ handle_nl_command_ack(
 {
     int *ret = arg;
     *ret = 0;
+    DBG("%d", *ret);
     return NL_STOP;
 }
 
 static
 int
-handle_nl_command_family(
+handle_nl_seq_check(
     struct nl_msg *msg,
     void *arg)
 {
-    struct multicast_group *grp = arg;
-    struct nlattr *tb[CTRL_ATTR_MAX + 1];
-    struct genlmsghdr *gnlh = nlmsg_data(nlmsg_hdr(msg));
-    struct nlattr *mcgrp;
-    int rem_mcgrp;
-
-    nla_parse(tb, CTRL_ATTR_MAX, genlmsg_attrdata(gnlh, 0), genlmsg_attrlen(gnlh, 0), NULL);
-
-    if (!tb[CTRL_ATTR_MCAST_GROUPS]) return NL_SKIP;
+    DBG("");
 
-    nla_for_each_nested(mcgrp, tb[CTRL_ATTR_MCAST_GROUPS], rem_mcgrp) {
-        struct nlattr *tb_mcgrp[CTRL_ATTR_MCAST_GRP_MAX + 1];
+    return NL_OK;
+}
 
-        nla_parse(tb_mcgrp, CTRL_ATTR_MCAST_GRP_MAX, nla_data(mcgrp), nla_len(mcgrp), NULL);
+static
+gboolean
+suspend_plugin_netlink_handler()
+{
+    struct nl_cb *cb;
+    int res = 0;
+    int err = 0;
 
-        if (!tb_mcgrp[CTRL_ATTR_MCAST_GRP_NAME] || !tb_mcgrp[CTRL_ATTR_MCAST_GRP_ID]) {
-            continue;
-        }
+    cb = nl_cb_alloc(NL_CB_VERBOSE);
+    if (!cb) {
+        connman_error("%s: failed to allocate netlink callbacks", __func__);
+        return FALSE;
+    }
 
-        if (strncmp(nla_data(tb_mcgrp[CTRL_ATTR_MCAST_GRP_NAME]), grp->group,
-                nla_len(tb_mcgrp[CTRL_ATTR_MCAST_GRP_NAME]))) {
-            continue;
+    err = 1;
+    nl_cb_err(cb, NL_CB_CUSTOM, handle_nl_command_error, &err);
+    nl_cb_set(cb, NL_CB_VALID, NL_CB_CUSTOM, handle_nl_command_valid, &err);
+    nl_cb_set(cb, NL_CB_FINISH, NL_CB_CUSTOM, handle_nl_command_finished, &err);
+    nl_cb_set(cb, NL_CB_ACK, NL_CB_CUSTOM, handle_nl_command_ack, &err);
+    nl_cb_set(cb, NL_CB_SEQ_CHECK, NL_CB_CUSTOM, handle_nl_seq_check, &err);
+
+    while (err == 1) {
+        DBG("waiting until nl testmode command has been processed\n");
+        res = nl_recvmsgs(nl_socket, cb);
```

Comment:
Since we are processing messages only after sending requests, is it possible that we might (initially?) get unrelated data in?

Is it possible that we do not receive the expected results? Can this turn into forever wait?

Reply:
@spiiroin that was the initial issue with this

with the changes it shouldn't happen - and wpa_supplicant is using the same approach in send_and_recv